AZFP parser error catch to handle case of raw file with no temperature

emiliom opened the pull request:
Addresses remaining issue in #198, where an AZFP raw file without temperature data leads to a failure during `open_raw`. As diagnosed there, the failure comes about in `parser_azfp.compute_temp` due to invalid `R` values passed to `math.log`. A test has been added that will catch cases of the `R` denominator being <= 0 and return `np.nan` instead.

This has been tested with an AZFP file (and corresponding XML) provided by @dsmossman. 

Remaining tasks:
- [ ] Add AZFP test file to our test data
- [ ] Add a test

leewujung commented on echopype/tests/convert/test_convert_azfp.py:
```diff
@@ -161,3 +161,17 @@ def test_convert_azfp_01a_different_ranges(azfp_path):
+
+def test_convert_azfp_01a_notemperature(azfp_path):
+    """Test converting file with no temperature data."""
+    azfp_01a_path = str(azfp_path.joinpath('rutgers_glider_notemperature', '22052500.01A'))
+    azfp_xml_path = str(azfp_path.joinpath('rutgers_glider_notemperature', '22052501.XML'))
```
question: wouldn't the code work without the use of `str`?

emiliom replied:
I started the new test code by copying and pasting one of the existing tests in that module. Yes, `str` is unnecessary, but its use is also present in that context in every test in the module. I'd rather deal with that in the context of a broader clean up.

But feel free to push a commit that cleans this up :wink:

leewujung replied:
haha, I am going to push back and say: how about clean up just this test file? It's unclear to me when a broader clean up will happen, so might as well do small bits first.

emiliom replied:
Addressed in a new commit.

leewujung commented on echopype/convert/parse_azfp.py:
```diff
@@ -112,9 +112,14 @@ def compute_temp(counts):
             and the counts from ancillary
             """
             v_in = 2.5 * (counts / 65535)
+            # Occurs when temperature sensor is not active (eg, glider deployments)
+            if self.parameters["kc"] - v_in <= 0:
+                return np.nan
+
```
`self.parameters["kc"] - v_in <= 0` is fairly arbitrary since v_in is essentially trying to get some sensor voltage values.

Since in the case of the glider, all temperature-related parameters in `self.parameters` are 0, it is likely safer to test for these cases?

Or, alternatively, test for the case when `R<=0` so that we don't go into log(negative values).

emiliom replied:
Are we certain that the relevant temperature parameters (`ka`, `kb`, `kc`) will always be 0 in circumstances like the Rutgers glider where the temperature sensor is turned off (or not deployed?). Do you know if it's documented that that will always be the case? Assuming that is true, can we reduce the 0 tests to just one of the parameters, to minimize the footprint of the test? 

BTW, `v_in` values are small (and apparently constant) but non-zero. See https://github.com/OSOceanAcoustics/echopype/issues/198#issuecomment-791560031

I'll comment on `compute_tilt` below your comment.

emiliom replied:
Addressed in a new commit, for both `compute_temp` and `compute_tilt`.

leewujung reviewed (changes requested):
@emiliom : I have a suggestion re the statement you used to handle the temperature conversion error.

Also, while you are on it, I think we should change the tilt calculations too. All tilt-related parameters in `self.parameters` are also all 0 in the case of the glider, likely due to the same reason that the glider is recording the attitude separately just like the temperature. I suggest that in `compute_tilt` we check for the parameter values, and if they are all 0, return NaN (instead of 0 as it currently returns), so that users do not think that they are actually getting tilt=0 data.

emiliom commented:
> Also, while you are on it, I think we should change the tilt calculations too. All tilt-related parameters in `self.parameters` are also all 0 in the case of the glider, likely due to the same reason that the glider is recording the attitude separately just like the temperature. I suggest that in `compute_tilt` we check for the parameter values, and if they are all 0, return NaN (instead of 0 as it currently returns), so that users do not think that they are actually getting tilt=0 data.

Makes sense. I'll take a closer look. I think it'll be more efficient, computationally, to do these tests outside the `compute_temp` and `compute_tilt` functions. I'll submit a commit with an implementation that you can review.

leewujung commented:
As an overall comment: I am about to write AZFP an email so I'll include the questions about whether these values are always zero in it!

emiliom commented:
New commits:
- Refactored the `compute_temp` test for invalid (= 0) temperature parameters.
- Added `compute_tilt` test for invalid (= 0) tilt parameters.

For both tests, I set the testing of actual parameter values outside the `compute_*` functions, so they happen only once. The functions are then passed a simple boolean.

Also, in the test module I replaced the use of `joinpath` with pathlib path concatenation.

leewujung commented:
Great, thanks @emiliom! I'll comment on this once we hear from ASL and if the 0s are indeed the default behavior, we can merge this.

Also, in looking at the code, I had the urge to move the `compute_*` functions to be private functions outside of `parse_raw` to make things more readable, and some location of the constants. Do you prefer for those to be in a separate PR?

emiliom commented:
Pushed new commit to add test checks for `tilt_x` and `tilt_y` similar to the one for `temperature`.

emiliom commented:
> Also, in looking at the code, I had the urge to move the `compute_*` functions to be private functions outside of `parse_raw` to make things more readable, and some location of the constants. Do you prefer for those to be in a separate PR?

You can submit a PR against my PR. I'll take a quick look before merging.

leewujung commented:
@emiliom : the movements end up to be a much broader job than I originally thought (it's too fun to refactor!), so I think I'll submit a separate PR for that...

leewujung commented:
I haven't heard from AZFP. I think we can merge this tomorrow if I don't hear back by then, so that this and #1024 can both get merged. We have a few small fixes in v0.7.1 and it'll be good to release by end of this week so that those changes get to actual package.

emiliom commented:
Per email from ASL Environmental to @leewujung, it's been confirmed that the temperature (and tilt?) parameters in the XML file are set to zero when the sensor is deactivated.

So, we can merge this PR as is, right? Then, PR #1024 can be merged, after updating it to reflect this PR.

leewujung commented:
Yes, the values for temperature and tilt sensor coefficients will be 0 in glider-specific AZFP deployments.

Feel free to merge this, and I'll update #1024 and merge that too!
